test/spec/modules/appnexusAstBidAdapter_spec.js

```diff
@@ -1,194 +1,165 @@
+    it('exists and is a function', () => {
+      expect(adapter.callBids).to.exist.and.to.be.a('function');
+    });
+  });
 
-    beforeEach(() => {
-      xhr = sinon.useFakeXMLHttpRequest();
-      requests = [];
-      xhr.onCreate = request => requests.push(request);
+  describe('isBidRequestValid', () => {
+    let bid = {
+      'bidder': 'appnexusAst',
+      'params': {
+        'placementId': '10433394'
+      },
+      'adUnitCode': 'adunit-code',
+      'sizes': [[300, 250], [300, 600]],
+      'bidId': '30b31c1838de1e',
+      'bidderRequestId': '22edbae2733bf6',
+      'auctionId': '1d1a030790a475',
+    };
+
+    it('should return true when required params found', () => {
+      expect(spec.isBidRequestValid(bid)).to.equal(true);
     });
 
-    afterEach(() => xhr.restore());
+    it('should return true when required params found', () => {
+      let bid = Object.assign({}, bid);
+      delete bid.params;
+      bid.params = {
+        'member': '1234',
+        'invCode': 'ABCD'
+      };
 
-    it('exists and is a function', () => {
-      expect(adapter.callBids).to.exist.and.to.be.a('function');
+      expect(spec.isBidRequestValid(bid)).to.equal(true);
     });
 
-    it('requires paramters to make request', () => {
-      adapter.callBids({});
-      expect(requests).to.be.empty;
+    it('should return false when required params are not passed', () => {
+      let bid = Object.assign({}, bid);
+      delete bid.params;
+      bid.params = {
+        'placementId': 0
+      };
+      expect(spec.isBidRequestValid(bid)).to.equal(false);
     });
+  });
+
+  describe('buildRequests', () => {
+    let bidRequests = [
+      {
+        'bidder': 'appnexusAst',
+        'params': {
+          'placementId': '10433394'
+        },
+        'adUnitCode': 'adunit-code',
+        'sizes': [[300, 250], [300, 600]],
+        'bidId': '30b31c1838de1e',
+        'bidderRequestId': '22edbae2733bf6',
+        'auctionId': '1d1a030790a475',
+      }
+    ];
 
-    it('requires member && invCode', () => {
-      let backup = REQUEST.bids[0].params;
-      REQUEST.bids[0].params = {member: 1234};
-      adapter.callBids(REQUEST);
-      expect(requests).to.be.empty;
-      REQUEST.bids[0].params = backup;
+    it('should add source and verison to the tag', () => {
+      const request = spec.buildRequests(bidRequests);
+      const payload = JSON.parse(request.data);
+      expect(payload.sdk).to.exist;
+      expect(payload.sdk).to.deep.equal({
+        source: 'pbjs',
+        version: '$prebid.version$'
+      });
     });
 
-    it('attaches valid video params to the tag', () => {
-      REQUEST.bids[0].params.video = {
+    it('sends bid request to ENDPOINT via POST', () => {
+      const request = spec.buildRequests(bidRequests);
+      expect(request.url).to.equal(ENDPOINT);
+      expect(request.method).to.equal('POST');
+    });
+
+    it('should attach valid video params to the tag', () => {
+      bidRequests[0].params.video = {
```

Comment:
This is a side-effect. If the test fails, the `delete` line at the end will never be executed, which causes cascading failures.

You've got to either copy the object before mutating, or move this to a `beforeEach` and `afterEach` block. Either one is fine... but for what it's worth, I tend to find the first one easier. `JSON.parse(JSON.stringify(obj))` is an inefficient, but concise way to make a deep copy.

The same goes for some of these other tests as well.

Reply:
Done